### PR TITLE
Add OEM Heat Pumps

### PR DESCRIPTION
**Pull request by mikivee**
## Description

Add 4 OEM heat pumps to the training set along with all of the other upgrades. 3 of these heat pump were added in #38 as as part of an experiment, but we are not retraining the full surrogate model with these OEM heat pumps. This required the following changes:
* OEM Heat pumps and performance curves:
     * Add one new OEM ducted heat pump (15.08), which required updating dml-utils, thus changing all the requirements files
     * Add performance curve parameters for all NREL heat pumps
     * Add new upgrades and performance curve params to feature upgrade unit test cases
     * NIT: small renaming of performance curves from `*capacity*` ->  `*capacity_retention*` as requested [here](https://github.com/rewiringamerica/surrogate_modeling/pull/38#discussion_r2044874943).  
* Slightly changing scope of universe of houses in training set:
    * Removing homes without heating from the training set since these have terrible performance and so we arent doing inference on these anyway and they are just adding noise to training
    * Temporarily removing homes with heat pumps in baseline since we don't have performance curves for these yet, @mfathollahzadeh will pull these eventually
* Decreased patience from 15 to 12 epoch to reduce overfitting, which we were starting to see signs of
* Change sample weighting to increase the weight of baseline samples relative to upgrade samples. As we continue to add more upgrades, baseline samples make up a smaller and smaller proportion of the training set, even though they make up 50% of the savings calculations. Further, they tend to have a much higher energy consumption, and as the model sees more and more upgrades, it tends to start underpredicting energy consumption for baseline. Each time we have added upgrades lately, we have seen overall lower test error (as we would expect with adding more training data), but higher savings error since the baseline error would increase, which thus increasing savings error. This sample weighting seems to resolve this issue, and the savings error has now decreased for every upgrade compared to the model current in production (see figure below). 
* New [exploratory notebook](https://github.com/rewiringamerica/surrogate_modeling/blob/6c91079e282d205ca4ec6b3ed6f837acd1257a17/notebooks/Prediction%20%20Diagnostics.ipynb) for diagnosing model performance, particularly related to model bias. This notebook helped identify the need for sample weighting described in the bullet above. 
* Evaluation model artifacts: updated with results from latest training run

## Results

Lower savings error across the board due to a larger training set and the upweighted baseline samples as described above. New OEM heat pumps have comparable error to NREL heat pumps , with ~7-9% Median APE. 

image

### CODE REVIEW

**Review by tsgra — COMMENTED** [review]
Super cool - great innovation to upweight the baseline samples. 

I saw in your commits that you're removing tests that require spark. It is possible to include these - see this [fixture](https://github.com/rewiringamerica/upfront-costs/blob/d9bdceb4b426f9327a754543368e23c484278cfe/tests/test_process_tech_data.py#L18) in some tests for upfront costs. I think you'd just have to make sure that pyspark is in your pyproject.toml so the poetry install during the github action includes it. It can be finicky and I don't want to cause headaches, but if you think the tests are important could be worth including.

File: src/datagen.py [hunk 1]
```diff
@@ -99,6 +108,21 @@ def __init__(
...
+        Parameters
+        ----------
+        - target_proportion (float): The targeted proportional weight of the baseline samples in the training set. Defaults to .5, since they make up half of every savings calculation.
+
+        Returns
+        -------
+        - float: baseline sample weight
+        """
+
+        return target_proportion * len(self.data_params["upgrade_ids"]) / len(self.baseline_ids)
```
Comment (tsgra) [inline src/datagen.py:124]: Potentially blocking: One thing I did notice here is that if there are no baseline ids, you will get division by 0 which would throw an error I'm assuming. Perhaps the likelihood of a batch have no baseline samples is so unlikely that this will never happen? Or it is explicitly prevented from happening?

File: src/datagen.py [hunk 2]
```diff
@@ -82,6 +88,9 @@ def __init__(
         self.building_features = building_features
         self.weather_features = weather_features
         self.targets = targets
+        # get the upgrade ids of the of baseline upgrades in the training set
+        self.baseline_ids = [x for x in self.data_params["upgrade_ids"] for b in baseline_building_sets if x == b[1]]
+        self.baseline_weight = self.get_baseline_sample_weight(target_proportion=baseline_weight_target_proportion)
```
Comment (tsgra) [inline src/datagen.py:93]: Potentially blocking: Does including this in the `__init__` method mean that the baseline sample weight is calculated based on the relative proportions of baseline and upgrade samples in the whole training dataset? If so, should it be moved to the `__getitem__` method so that the target proportion is achieved _for each batch_ and the composition of the loss is not mismatched with the composition of the samples?

**Review by tsgra — CHANGES_REQUESTED** [review]
Adding a few more potentially blocking comments around the sample weighting.

**Comment by mikivee** [comment]
> Super cool - great innovation to upweight the baseline samples.
> 
> I saw in your commits that you're removing tests that require spark. It is possible to include these - see this [fixture](https://github.com/rewiringamerica/upfront-costs/blob/d9bdceb4b426f9327a754543368e23c484278cfe/tests/test_process_tech_data.py#L18) in some tests for upfront costs. I think you'd just have to make sure that pyspark is in your pyproject.toml so the poetry install during the github action includes it. It can be finicky and I don't want to cause headaches, but if you think the tests are important could be worth including.

Yes I know that is a thing, and I would really like to have these tests run in github, but my concern is that since we are using the radbcluster to install all the same packages as databricks (which includes spark), I won't be able to cleanly include this in the pyproject.toml. Have you successfully done this for a repo that depends on radbcluster version 14.3?

File: src/datagen.py [hunk 3]
```diff
@@ -62,13 +65,16 @@ def __init__(
         targets: List[str] = None,
         batch_size: int = 256,
         dtype: np.dtype = np.float32,
+        baseline_weight_target_proportion=0.5,
```
[the remark this answers is not in the document]
Comment (mikivee) [reply src/datagen.py:68]: Good question! No i haven't. My thought was that the baseline samples are half of every savings calculation and so .5 would be the good target proportion. I adjusted it down from there to .3 (frankly pretty arbitrarily) because I wanted to see how it performed with a less extreme re-weighting. To be honest, I'm hesitant to spend a lot of time/effort on tuning this because incorporating savings into the loss project that hopefully you will do, will remove the need for this reweighting at all, since baselines samples will implicitly be overweighted in the savings loss. This is simply an stopgap solution to let us limp along until then. However, I imagine once we get to that, you will want a way to explore over a grid of weights for the savings vs baseline vs upgrade error in the loss, which will essentially get at the same thing. 

Wrt to upfront costs, if you are using `RandomizedSearchCV` (which I think you are), i think that since [`sample_weight` is a valid arg for the `fit()` function](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html#sklearn.model_selection.RandomizedSearchCV.fit), that the way to pass in a grid of sample weights is to pass[ it to `fit_params={'sample_weight': sample_weights}` as an arg to RandomizedSearchCV()](https://scikit-learn.org/0.16/modules/generated/sklearn.grid_search.RandomizedSearchCV.html). Did you already try that?

File: src/datagen.py (hunk 1 above)
Comment (mikivee) [reply src/datagen.py:124]: Oh good call. In fact, the `load_data()` function will fail as well if you don't have any baseline samples in your training set. Let me make that explicit in that function, and then add better handling here.

File: src/datagen.py (hunk 2 above)
Comment (mikivee) [reply src/datagen.py:93]: Oh good point I hadn't considered that! I was more thinking that we wanted the overall distribution to be, and that that general weighting scheme would inform the relative weights for each batch. My sense is that having it at the training set level is better BUT that it would certainly make sense to test both options if this was was our long term solution. However given my comment [here](https://github.com/rewiringamerica/surrogate_modeling/pull/39#discussion_r2058301157) about this being an interim solution, I'm again hesitant to sink a lot of time into this and retraining. I think I'd like to leave a note comment inine to document this idea and alternative choice tho, just in case this sample weighting does actually persist. Does this make sense to you?

Comment (tsgra) [reply src/datagen.py:93]: Sure!

**Review by tsgra — APPROVED** [review]
Thanks for addressing comments.
